include/cuco/bloom_filter.cuh

```diff
@@ -0,0 +1,393 @@
+   * <tt>std::is_convertible<std::iterator_traits<InputIt>::value_type,
+   * bloom_filter<K>::key_type></tt> is `true`
+   * @tparam StencilIt Device-accessible random-access iterator whose `value_type` is
+   * convertible to Predicate's argument type
+   * @tparam Predicate Unary predicate callable whose return type must be convertible to `bool` and
+   * argument type is convertible from <tt>std::iterator_traits<StencilIt>::value_type</tt>
+   * @tparam OutputIt Device-accessible output iterator assignable from `bool`
+   *
+   * @param first Beginning of the sequence of keys
+   * @param last End of the sequence of keys
+   * @param stencil Beginning of the stencil sequence
+   * @param pred Predicate to test on every element in the range `[stencil, stencil +
+   * std::distance(first, last))`
+   * @param output_begin Beginning of the sequence of booleans for the presence of each key
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <class InputIt, class StencilIt, class Predicate, class OutputIt>
+  __host__ void test_if_async(InputIt first,
+                              InputIt last,
+                              StencilIt stencil,
+                              Predicate pred,
+                              OutputIt output_begin,
+                              cuda::stream_ref stream = {}) const noexcept;
+
+  /**
+   * @brief Gets a pointer to the underlying filter storage.
+   *
+   * @return Pointer to the underlying filter storage
+   */
+  [[nodiscard]] __host__ word_type* data() noexcept;
+
+  /**
+   * @brief Gets a pointer to the underlying filter storage.
+   *
+   * @return Pointer to the underlying filter storage
+   */
+  [[nodiscard]] __host__ word_type const* data() const noexcept;
+
+  /**
+   * @brief Gets the number of sub-filter blocks.
+   *
+   * @return Number of sub-filter blocks
+   */
+  [[nodiscard]] __host__ extent_type block_extent() const noexcept;
+
+  /**
+   * @brief Gets the number of bits in a key's fingerprint
+   *
+   * @return The number of fingerprint bits
+   */
+  [[nodiscard]] __host__ hasher hash_function() const noexcept;
+
+  /**
+   * @brief Gets the function used to hash keys
+   *
+   * @return The function used to hash keys
+   */
+  [[nodiscard]] __host__ std::uint32_t pattern_bits() const noexcept;
+
+  /**
+   * @brief Gets the allocator.
+   *
+   * @return The allocator
+   */
+  [[nodiscard]] __host__ allocator_type allocator() const noexcept;
+
+  /**
+   * @brief Get device ref.
+   *
+   * @return Device ref of the current `bloom_filter` object
+   */
+  [[nodiscard]] __host__ ref_type<> ref() const noexcept;
+
+ private:
+  allocator_type allocator_;  ///< Allocator used to allocate device-accessible storage
+  std::unique_ptr<word_type, detail::custom_deleter<std::size_t, allocator_type>>
+    data_;          ///< Storage of the current `bloom_filter` object
+  ref_type<> ref_;  ///< Device ref of the current `bloom_filter` object
```

Comment:
Does this need to be a member? I'm wondering if this will make the container not thread-safe.

Reply:
`ref_` holds all the members that are needed to create a new ref, e.g. `pattern_bits`, `num_blocks`, etc.

This should be thread-safe since we never mutate the state of `ref_` after its creation.